rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,490 @@
+limited to X25519 (recommended by the spec), Ed25519, and RSA. Support for
+other algorithms would either have to be requested from the upstream or
+manually implemented as a custom plugin. The algorithms employed by `age` are
+not currently compatible with FIPS, which means configuring encrypted sessions
+while in FIPS mode will result in failed startup.
+
+Below is a high level diagram showing how `age` encryption and decryption work:
+![age diagram](assets/0127-age-high-level.png)
+
+### Config Changes
+
+Encrypted session recording is a feature of the auth service and can be enabled
+through the `session_recording_config` resource.
+```yaml
+# session_recording_config.yml
+kind: session_recording_config
+version: v2
+spec:
+  encryption:
+    enabled: true
+```
+HSM integration is facilitated through the existing configuration
+options for setting up an HSM backed CA keystore through pkcs#11. Example
+configuration found [here](https://goteleport.com/docs/admin-guides/deploy-a-cluster/hsm/#step-25-configure-teleport).
+
+### Protobuf Changes
+```proto
+// api/proto/teleport/legacy/types/types.proto
+
+// EncryptionKeyPair is a keypair used for encrypting and decrypting data.
+message EncryptionKeyPair {
+  // PublicKey is the public encryption key.
+  bytes PublicKey = 1 [(gogoproto.jsontag) = "public_key,omitempty"];
+  // PrivateKey is the private decryption key.
+  bytes PrivateKey = 2 [(gogoproto.jsontag) = "private_key,omitempty"];
+  // PrivateKeyType is the type of the PrivateKey.
+  PrivateKeyType PrivateKeyType = 3 [(gogoproto.jsontag) = "private_key_type,omitempty"];
+}
+
+// WrappedKey wraps a PrivateKey using an asymmetric keypair.
+message WrappedKey {
+  // WrappingPair is the asymmetric keypair used to wrap the private key.
+  // Expected to be RSA.
+  EncryptionKeyPair WrappingPair = 1;
+  // WrappedPair is the asymmetric keypair used with age to encrypt and decrypt
+  // filekeys.
+  EncryptionKeyPair WrappedPair = 2;
+  // Signals that a key should be rotated.
+  bool Rotate = 3 [(gogoproto.jsontag) = "rotate"];
+}
+
+// EncryptionKeySet contains the list of active and rotated WrappedKeys for a
+// given usage.
+message EncryptionKeySet {
+  // ActiveKeys is a list of active, wrapped X25519 private keys. There should
+  // be at most one wrapped key per auth server using the
+  // SessionRecordingConfigV2 resource unless keys are being rotated.
+  repeated WrappedKey ActiveKeys = 1 [
+    (gogoproto.jsontag) = "active_keys"
+  ];
+  // RotatedKeys is a list of wrapped private keys that have been rotated.
+  // These are kept to decrypt historical encrypted session recordings.
+  repeated WrappedKey RotatedKeys = 2 [
+    (gogoproto.jsontag) = "rotated_keys"
+  ];
+}
+// SessionRecordingConfigStatusV2 contains all of the current and rotated keys
+// used for encrypted session recording.
+message SessionRecordingConfigStatusV2 {
+  EncryptionKeySet KeySet = 1 [(gogoproto.jsontag) = "keyset"]
+}
```

Comment:
This is stored in the backend, likely as a resource, probably in json, each wrapped key is going to be something like 2KiB and we only have like 400KiB to play with in a single backend item - using a single item has the potential to be _very_ problematic after a cluster has existed for a while.

Reply:
Do we have any existing features that split across arbitrary numbers of backend items? Since we can't really throw out rotated keys, I agree this will definitely be a problem for someone at some point

Reply:
Since I split this into new resources, the base path would now be `/recording_encryption/rotated_keys/`. I'd probably just take the hex encoded sha256 of the public key as the fingerprint.